pyomo/core/util.py

```diff
@@ -0,0 +1,278 @@
+    """
+    A utility function to compute the product of a list of terms.
+
+    Args:
+        terms (list): A list of terms that are multiplied together.
+
+    Returns:
+        The value of the product, which may be a Pyomo expression object.
+    """
+    ans = 1
+    for term in terms:
+        ans *= term
+    return ans
+
+
+def quicksum(args, start=0, linear=None):
+    """
+    A utility function to compute a sum of Pyomo expressions.
+
+    The behavior of :func:`quicksum` is similar to the builtin :func:`sum`
+    function, but this function generates a more compact Pyomo
+    expression.
+
+    .. note::
+        Is there a better name for this function?  quicksum() and sum() are very similar, which could make it 
+        difficult to see whether this function is being used.  But that similarity means that changes
+        from standard Python expressions are quite similar.
+
+    Args:
+        args: A generator for terms in the sum.
+
+        start: A value that is initializes the sum.  If
+            this value is not a numeric constant, then the += 
+            operator is used to add terms to this object.
+            Defaults to zero.
+
+        linear: If :attr:`start` is not a numeric constant, then this 
+            option is ignored.  Otherwise, this value indicates
+            whether the terms in the sum are linear.  If the value
+            is :const:`False`, then the terms are
+            treated as nonlinear, and if :const:`True`, then
+            the terms are treated as linear.  Default is
+            :const:`None`, which indicates that the first term
+            in the :attr:`args` is used to determine this value.
+
+    Returns:
+        The value of the sum, which may be a Pyomo expression object.
+    """
+    #
+    # If we're starting with a numeric value, then 
+    # create a new nonlinear sum expression but 
+    # return a static version to the user.
+    #
+    if start.__class__ in native_numeric_types:
+        if linear is None:
+            #
+            # Get the first term, which we will test for linearity
+            #
+            first = next(args, None)
+            if first is None:
+                return start
+            #
+            # Check if the first term is linear, and if so return the terms
+            #
+            linear, terms = decompose_term(first)
+            #
+            # Right now Pyomo5 expressions can only handle single linear
+            # terms.
+            #
+            # Also, we treat linear expressions as nonlinear if the constant
+            # term is not a native numeric type.  Otherwise, large summation
+            # objects are created for the constant term.
+            #
+            if linear:
+                nvar=0
+                for term in terms:
+                    c,v = term
+                    if not v is None:
+                        nvar += 1
+                    elif not c.__class__ in native_numeric_types:
```

Comment:
Should this check be changed to `elif c.is_potentially_variable():`? The example I'm using is
```python
import pyomo.kernel as pmo
from pyomo.environ import quicksum

m = pmo.block()
m.d = pmo.data_expression(1.0)

print(quicksum(m.d for i in range(10)).__class__.__name__)
```
Which seems to generate the general "nonlinear" `ViewSumExpression`.

Reply:
Actually, I suppose it needs to be `and`-ed with the current check so that simple numeric types don't result in an error:

`(c.__class__ not in native_numeric_types) and c.is_potentially_variable():`

Reply:
@ghackebeil We need to move this example into a separate ticket.  I think the answer is "no", but then I don't really understand what a data_expression(1.0) object is.

Reply:
@whart222: The `data_expression` component is just an example of something that is NPV and mutable. You could replace it with a mutable parameter in the above and it would capture the same behavior. I've opened #431 and modified the example there.